### PR TITLE
Optimisations: use bytes.IndexByte() + delay unescaping (#264)

### PR DESCRIPTION
**Pull request by kirillx**
There are 2 issues with current implementation (addresses issue #264 ):

1. It performs a plain byte to byte loop to find string boundaries and
perform unescaping. Replace this with bytes.IndexByte() implementation.

2. It performs unescaping of string values even when this is not really
needed, e.g. for members which are absent in target data structure.

This patch fixes both issues and results in ~12% faster BenchmarkEJ_Unmarshal_M-8,
plus number of allocations goes down from 128 to 52 (!):

```
benchmark                                     old MB/s     new MB/s     speedup
BenchmarkEJ_Unmarshal_M-8                     317.99       356.27       1.12x
BenchmarkEJ_Unmarshal_S-8                     142.19       139.77       0.98x

benchmark                                     old allocs     new allocs     delta
BenchmarkEJ_Unmarshal_M-8                     128            52             -59.38%
BenchmarkEJ_Unmarshal_S-8                     3              3              +0.00%
```

The rest of std benchmarks are w/o changes. The benchmark in the original issue became about 1.5x faster.

NOTE: performance is even more improved if disable unescaping of member names, ints, base64 bytes using a new -disable_members_unescape option:

```
benchmark                                     old MB/s     new MB/s     speedup
BenchmarkEJ_Unmarshal_M-8                     317.99       406.93       1.28x
BenchmarkEJ_Unmarshal_S-8                     142.19       150.85       1.06x
```

### CODE REVIEW

File: jlexer/lexer.go [hunk 1]
```diff
@@ -240,23 +242,50 @@ func (r *Lexer) fetchNumber() {
...
+		length += idx + 1
+		data = data[idx+1:]
+	}
+}
+
+// unescapeStringToken performs unescaping of string token.
+// if no escaping is needed, original string is returned, otherwise - a new one allocated
+func (r *Lexer) unescapeStringToken() (err error) {
+	data := r.token.byteValue
+	var unescapedData []byte
```
Comment (GoWebProd) [inline jlexer/lexer.go]: I think that you need to preallocate array with cap = len(data)

Comment (kirillx) [reply jlexer/lexer.go]: Hmmm, I can't explain it, but while preallocating makes number of allocations in benchmark lower (52 -> 46), the performance gain becomes less - 6% vs. 12% w/o preallocating...

**Comment by kirillx** [comment]
@GoWebProd What do you think about unescaping member names?
imho it's useless and can be safely removed - improves performance for another 12%.

**Comment by GoWebProd** [comment]
> @GoWebProd What do you think about unescaping member names?
> imho it's useless and can be safely removed - improves performance for another 12%.

I'm agree, this is really useless operation

**Comment by GoWebProd** [comment]
But in some cases this is needed, for example: https://play.golang.org/p/lX-d1iFGhFh

**Comment by rvasily** [comment]
> @GoWebProd What do you think about unescaping member names?
> imho it's useless and can be safely removed - improves performance for another 12%.

It can be only under the config option ( disabled by default ). Otherwise it will be very fragile.

File: jlexer/lexer.go [hunk 2]
```diff
@@ -273,6 +274,12 @@ func (r *Lexer) unescapeStringToken() (err error) {
 			r.errParse(err.Error())
 			return err
 		}
+
+		if !wasEscaped {
```
[the remark this answers is not in the document]
Comment (kirillx) [reply jlexer/lexer.go]: done

**Comment by kirillx** [comment]
> > @GoWebProd What do you think about unescaping member names?
> > imho it's useless and can be safely removed - improves performance for another 12%.
> 
> It can be only under the config option ( disabled by default ). Otherwise it will be very fragile. We must support all corner cases

@rvasily -disable_members_unescape option, see PR comments

**Comment by GoWebProd** [comment]
@kirillx can you write additional test with disable_members_unescape?

**Comment by kirillx** [comment]
> @kirillx can you write additional test with disable_members_unescape?

please suggest what tests cases you want to see, will add.

**Comment by GoWebProd** [comment]
> > @kirillx can you write additional test with disable_members_unescape?
> 
> please suggest what tests cases you want to see, will add.

https://play.golang.org/p/lX-d1iFGhFh for example, that this code works without flag and don't work with flag. And please, resolve conflicts. Thanks.

**Comment by kirillx** [comment]
@GoWebProd done, tests added
